pkg/datarepair/queue.go

```diff
@@ -3,40 +3,95 @@
 
 package datarepair
 
-import "storj.io/storj/pkg/pb"
+import (
+	"encoding/binary"
+	"time"
+	"sync"
+	"github.com/golang/protobuf/proto"
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/storage"
+	"storj.io/storj/storage/redis"
+)
 
 //Queue implements the RepairQueue interface
 type Queue struct {
-	//redis db of repair segments
-	//offline nodes?
+	DB storage.KeyValueStore
+	mutex *sync.Mutex
 }
 
-//NewQueue creates a new data repair queue
-func NewQueue() {
+var (
+	queueError = errs.Class("data repair queue error")
+)
 
+//NewQueue returns a pointer to a new Queue instance with an initialized connection to Redis
+func NewQueue(address, password string, db int) (*Queue, error) {
+	rc, err := redis.NewClient(address, password, db)
+	if err != nil {
+		return nil, err
+	}
+	return &Queue{
+		DB: rc,
+		mutex: &sync.Mutex{},
+	}, nil
 }
 
 //Add adds a repair segment to the queue
-func (q Queue) Add(qi *pb.InjuredSegment) error {
-	return nil
-}
-
-//AddAll adds a slice of repair segements to the queue
-func (q Queue) AddAll(qis []*pb.InjuredSegment) error {
-	return nil
+func (q *Queue) Add(qi *pb.InjuredSegment) (storage.Key, error) {
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+	dateTime := make([]byte, binary.MaxVarintLen64)
+	binary.PutVarint(dateTime, time.Now().UnixNano())
+	val, err := proto.Marshal(qi)
+	if err != nil {
+		return nil, queueError.New("error marshalling injured seg %s", err)
+	}
+	err = q.DB.Put(dateTime, val)
+	if err != nil {
+		return nil, queueError.New("error adding injured seg to queue %s", err)
+	}
+	return dateTime, nil
 }
 
 //Remove removes a repair segment from the queue
-func (q Queue) Remove(qi *pb.InjuredSegment) error {
+func (q *Queue) Remove(date storage.Key) error {
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+	err := q.DB.Delete(date)
+	if err != nil {
+		return queueError.New("error removing injured seg %s", err)
+	}
 	return nil
 }
 
 //GetNext returns the next repair segement from the queue
-func (q Queue) GetNext() pb.InjuredSegment {
-	return pb.InjuredSegment{}
+func (q *Queue) GetNext() (storage.Key, pb.InjuredSegment, error) {
```

Comment:
I wonder why other methods take a pointer to `pb.InjuredSegment` as param, but this one does not return a pointer too?